platform/stm32f0/drivers/componolit-runtime-drivers-rcc.adb

```diff
@@ -1,28 +1,28 @@
 
 package body Componolit.Runtime.Drivers.RCC with
-   SPARK_Mode
+   SPARK_Mode,
+   Refined_State => (RCC_State => Reg)
 is
+
    use type SSE.Integer_Address;
 
+   Reg : Register with
+      Address => SSE.To_Address (RCC_Base + AHB_EN_Offset),
+      Import;
+
    procedure Set (Clk    : Clock;
                   Enable : Boolean)
    is
-      Reg :   Register with
-         Address => SSE.To_Address (RCC_Base + AHB_EN_Offset),
-         Import;
       Cache : Register := Reg;
    begin
-      Cache (Clk'Enum_Rep) := (if Enable then 1 else 0);
+      Cache (Clock_Bit (Clk)) := (if Enable then 1 else 0);
```

Comment:
Do you really mean `Cache`? Or rather something like `Tmp`? The name confused me (could be related to CPU cache handling).

Reply:
I'm not 100% if it is needed at all. The problem here is that we assign a bit to a packed array of bits. Since access is at least byte-wise I read the register, change it and write it back. Maybe just assigning the bit to the register would cause the compiler to do the same?

Reply:
The compiler does that for you. If writing the register has no effects (i.e. no `Effective_Writes`), this should be no problem. Is it on purpose that `Reg` has no `Async_*` annotation?

Maybe we should also consider making `Register` an array of `Boolean` (instead of `Bit`). With this you could assign `Enable` directly.

Reply:
> If writing the register has no effects (i.e. no Effective_Writes), this should be no problem. Is it on purpose that Reg has no Async_* annotation?

You are right, it should have `Async_Readers` and `Effective_Writes`. How would effective writes be a problem in this case? I'm writing to the register anyway. I think effective reads would be more problematic but this is not the case here.

Reply:
But your intention is to write only one single bit. If what the compiler does is reading the register, masking in your bit and writing the 32 bit word back to the register, then unrelated bits that may be set would be written to the register again and may have an effect.

I guess hardware on which this is an issue would be problematic (but I don't think this is a real issue here)

Reply:
Adding `Async_Readers` and `Effective_Writes` to the configuration registers imposes `Volatile_Function` to the `Pin_Mode` function. The problem is that the result of this function isn't actually volatile. It isn't changing unless the configuration is changed (which is already modeled by the global state). Furthermore this prevents the function from being used in contracts.

Reply:
Is that also true when `Config_Registers` only has `Async_Readers`?

Would it even be semantically correct to model it as a plain package variable? I guess we'll get flow errors if you set the config without reading it in between.

Could we store the pin mode in a shadow package variable instead which is used for `Pin_Mode` only? `Config_Registers` would only be accessed when changing the configuration (in which case the shadow variable is also updated) ?

Reply:
Adding only `Async_Readers` also makes it volatile. Also `Effective_Writes` is a valid state description since the a configuration change might e.g. change the value of a pin.

How would you model that shadow variable? I'm asking myself if we really can be sure that a pin is in a certain state. Currently this is only used as precondition in the nRF52 driver for the `Write` procedure. We could however define that `Write` is always allowed and the value is ignored if the Pin is set to input.

Reply:
> How would you model that shadow variable?

Just a plain package variable?

> I'm asking myself if we really can be sure that a pin is in a certain state. Currently this is only used as precondition in the nRF52 driver for the `Write` procedure.

We can only be sure if we set it. But isn't that the case anyway with the hardware?

> We could however define that `Write` is always allowed and the value is ignored if the Pin is set to input.

I don't like this idea as it would hide errors in your code where you forgot to set the correct output mode and then wonder why it does not have the desired effect. It think we should strive for a more robust interface.

Reply:
This would then be an overlay over the actual configuration that is "read only"? `Configure` would then have to modify the state of this variable, too. The initial configuration could probably be read in the elaboration code.

Reply:
It would be a *shadow* (not and overlay in the "address overlay" sense). `Configure` sets the `Direction` as well as the regular package variable `Shadow_Direction` (or a more suitable name).

When reading the code I'd think that `Configure` should have a postcondition `Configured (P, M)` and `Read` should have a precondition `Configured (P, Port_In)`, `Write` a precondition `Configured (P, Port_Out)` and `Pin_Mode` a precondition `Configured (P, Port_In) or Configured (P, Port_Out)`. If `Configured` is only established by `Configure` this should ensure that the shadow state is always initialized (without doing that in the elaboration code).

Reply:
In this case should we ever read the original configuration register (except for its initial value). We could work only on the shadow register and configure would just copy the whole shadow register to the application. Assuming the state never externally changes it should have no unforseen side effects (if the state externally changes the shadow register will be out of sync so it just should not happen). When we establish a post condition, would it make sense to make `Pin_Mode` ghost? In this case one cannot check for the configuration of a pin and has to configure it before using. I think enforcing configuration before use is something we want to have.

Reply:
Fixed in f9b9800c9ce35a149d3d6a5ffdea53e2fe7c33d8